File: examples/javascript/batch_archive.js
/**
 * Batch archive example - Archive multiple URLs efficiently.
 *
 * Usage:
 *   node batch_archive.js
 *
 * Requirements:
 *   npm install node-fetch
 */

const fetch = require('node-fetch');

/**
 * HTBase API client.
 */
class HTBaseClient {
  /**
   * Create HTBase client.
   *
   * @param {string} baseUrl - HTBase API base URL
   */
  constructor(baseUrl = 'http://localhost:8000') {
    this.baseUrl = baseUrl;
  }

  /**
   * Submit batch archive request.
   *
   * @param {Array<{url: string, id: string}>} items - Items to archive
   * @param {string} archiver - Archiver to use
   * @returns {Promise<string>} Task ID
   */
  async batchArchive(items, archiver = 'readability') {
    const endpoint = `${this.baseUrl}/api/batch/${archiver}`;

    const response = await fetch(endpoint, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ items }),
      timeout: 30000
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`HTTP ${response.status}: ${text}`);
    }

    const result = await response.json();
    return result.task_id;
  }

  /**
   * Get status of a batch task.
   *
   * @param {string} taskId - Task ID from batchArchive()
   * @returns {Promise<object>} Task status
   */
  async getTaskStatus(taskId) {
    const endpoint = `${this.baseUrl}/api/tasks/${taskId}`;

    const response = await fetch(endpoint, {
      timeout: 10000
    });

    if (!response.ok) {
      const text = await response.text();
      throw new Error(`HTTP ${response.status}: ${text}`);
    }

    return response.json();
  }

  /**
   * Wait for batch task to complete.
   *
   * @param {string} taskId - Task ID to wait for
   * @param {number} pollInterval - Seconds between status checks
   * @param {number} maxWait - Maximum seconds to wait
   * @returns {Promise<object>} Final task status
   */
  async waitForCompletion(taskId, pollInterval = 5, maxWait = 600) {
    const start = Date.now();

    while (Date.now() - start < maxWait * 1000) {
      const status = await this.getTaskStatus(taskId);

      // Check if completed
      if (status.status === 'success' || status.status === 'failed') {
        return status;
      }

      // Wait before next poll
      await new Promise(resolve => setTimeout(resolve, pollInterval * 1000));
    }

    throw new Error(`Task ${taskId} did not complete in ${maxWait}s`);
  }
}

/**
 * Main function - example usage.
 */
async function main() {
  // Initialize client
  const client = new HTBaseClient();

  // URLs to archive
  const urlsToArchive = [
    'https://example.com/article-1',
    'https://example.com/article-2',
    'https://example.com/article-3',
    'https://example.com/article-4',
    'https://example.com/article-5'
  ];

  // Prepare batch items
  const items = urlsToArchive.map((url, i) => ({
    url: url,
    id: `article-${i + 1}`
  }));

  console.log(`Batch archiving ${items.length} URLs...`);
  console.log('-'.repeat(50));

  try {
    // Submit batch request
    const taskId = await client.batchArchive(items, 'readability');
    console.log(`✅ Batch task created: ${taskId}`);

    // Wait for completion
    console.log('Waiting for completion...');
    const result = await client.waitForCompletion(taskId, 5);

    // Print results
    console.log('-'.repeat(50));
    console.log(`Batch status: ${result.status}`);
    console.log(`Total items: ${result.items.length}`);

    // Count successes and failures
    const successes = result.items.filter(item => item.status === 'success');
    const failures = result.items.filter(item => item.status === 'failed');

    console.log(`✅ Successful: ${successes.length}`);
    console.log(`❌ Failed: ${failures.length}`);

    // Print details
    if (successes.length > 0) {
      console.log('\nSuccessful archives:');
      successes.forEach(item => {
        console.log(`  - ${item.id}: ${item.saved_path}`);
      });
    }

    if (failures.length > 0) {
      console.log('\nFailed archives:');
      failures.forEach(item => {
        console.log(`  - ${item.id}: exit_code=${item.exit_code || 'unknown'}`);
      });
    }
  } catch (error) {
    console.error('❌ Error:', error.message);
  }
}

// Run if executed directly
if (require.main === module) {
  main().catch(console.error);
}

// Export for use as module
module.exports = { HTBaseClient };

File: examples/python/batch_archive.py
"""
Batch archive example - Archive multiple URLs efficiently.

Usage:
    python batch_archive.py

Requirements:
    pip install requests
"""

import requests
import time
from typing import List, Dict, Any


class HTBaseClient:
    """Client for HTBase API."""

    def __init__(self, base_url: str = "http://localhost:8000"):
        """
        Initialize HTBase client.

        Args:
            base_url: HTBase API base URL
        """
        self.base_url = base_url

    def batch_archive(
        self,
        items: List[Dict[str, str]],
        archiver: str = "readability"
    ) -> str:
        """
        Submit batch archive request.

        Args:
            items: List of dicts with 'url' and 'id' keys
            archiver: Archiver to use

        Returns:
            Task ID for tracking

        Raises:
            requests.exceptions.RequestException: If the request fails
        """
        endpoint = f"{self.base_url}/api/batch/{archiver}"

        response = requests.post(
            endpoint,
            json={"items": items},
            timeout=30
        )
        response.raise_for_status()

        result = response.json()
        return result["task_id"]

    def get_task_status(self, task_id: str) -> Dict[str, Any]:
        """
        Get status of a batch task.

        Args:
            task_id: Task ID from batch_archive()

        Returns:
            Dictionary with task status

        Raises:
            requests.exceptions.RequestException: If the request fails
        """
        endpoint = f"{self.base_url}/api/tasks/{task_id}"

        response = requests.get(endpoint, timeout=10)
        response.raise_for_status()

        return response.json()

    def wait_for_completion(
        self,
        task_id: str,
        poll_interval: int = 5,
        max_wait: int = 600
    ) -> Dict[str, Any]:
        """
        Wait for batch task to complete.

        Args:
            task_id: Task ID to wait for
            poll_interval: Seconds between status checks
            max_wait: Maximum seconds to wait

        Returns:
            Final task status

        Raises:
            TimeoutError: If task doesn't complete in max_wait seconds
        """
        start = time.time()
        while time.time() - start < max_wait:
            status = self.get_task_status(task_id)

            # Check if completed
            if status["status"] in ["success", "failed"]:
                return status

            # Wait before next poll
            time.sleep(poll_interval)

        raise TimeoutError(f"Task {task_id} did not complete in {max_wait}s")


def main():
    """Example usage."""
    # Initialize client
    client = HTBaseClient()

    # URLs to archive
    urls_to_archive = [
        "https://example.com/article-1",
        "https://example.com/article-2",
        "https://example.com/article-3",
        "https://example.com/article-4",
        "https://example.com/article-5",
    ]

    # Prepare batch items
    items = [
        {
            "url": url,
            "id": f"article-{i+1}"
        }
        for i, url in enumerate(urls_to_archive)
    ]

    print(f"Batch archiving {len(items)} URLs...")
    print("-" * 50)

    try:
        # Submit batch request
        task_id = client.batch_archive(items, archiver="readability")
        print(f"✅ Batch task created: {task_id}")

        # Wait for completion
        print("Waiting for completion...")
        result = client.wait_for_completion(task_id, poll_interval=5)

        # Print results
        print("-" * 50)
        print(f"Batch status: {result['status']}")
        print(f"Total items: {len(result['items'])}")

        # Count successes and failures
        successes = [item for item in result["items"] if item["status"] == "success"]
        failures = [item for item in result["items"] if item["status"] == "failed"]

        print(f"✅ Successful: {len(successes)}")
        print(f"❌ Failed: {len(failures)}")

        # Print details
        if successes:
            print("\nSuccessful archives:")
            for item in successes:
                print(f"  - {item['id']}: {item['saved_path']}")

        if failures:
            print("\nFailed archives:")
            for item in failures:
                print(f"  - {item['id']}: exit_code={item.get('exit_code')}")

    except TimeoutError as e:
        print(f"❌ {e}")

    except requests.exceptions.RequestException as e:
        print(f"❌ Request failed: {e}")


if __name__ == "__main__":
    main()
